pyensign/ensign.py

```diff
@@ -474,3 +489,203 @@ def _resolve_topic(self, topic):
+            return res
+
+        return wrapper
+
+    def wrap_async_generator(coro):
+        async def wrapper(*args, **kwargs):
+            global _client
+            if _client is not None:
+                async for res in coro(*args, **kwargs):
+                    yield res
+
+            # Set the global client for the duration of the generator
+            try:
+                async with Ensign(*auth_args, **auth_kwargs) as client:
+                    _client = client
+                    async for res in coro(*args, **kwargs):
+                        yield res
+            except Exception as e:
+                _client = None
+                raise e
+
+            _client = None
+
+        return wrapper
+
+    # Return either a coroutine or an async generator wrapper to match the marked
+    # function
+    def decorator(coro):
+        if inspect.iscoroutinefunction(coro):
+            return wrap_coroutine(coro)
+        elif inspect.isasyncgenfunction(coro):
+            return wrap_async_generator(coro)
+        else:
+            raise TypeError(
+                "decorated function must be a coroutine or async generator, got {}".format(
+                    type(coro)
+                )
+            )
+
+    return decorator
+
+
+def publish(topic, mimetype=None, encoder=None):
+    """
+    Decorator to mark a publish function. The return value of the function will be
+    published as an event to the topic. If the function is a generator, the events are
+    published as they are yielded from the generator. This method assumes that you have
+    already authenticated with Ensign using the `@authenticate` decorator.
+
+    Parameters
+    ----------
+    topic: str
+        The name or ID of the topic to publish events to.
+
+    mimetype: str or int (optional)
+        The mimetype indicating how the return value will be encoded. If not provided,
+        the mimetype is inferred from the return value of the function.
+
+    encoder: events.BytesEncoder (optional)
+        The encoder to use to encode the return value into bytes for the event payload.
+        If not provided, a best effort is made to encode the value, for most data types
+        this will be JSON.
+
+    Usage
+    -----
+    ```
+    # Publish a JSON event to the topic 'my-topic'
+    @publish("my-topic")
+    async def my_function():
+        return {"foo": "bar"}
+
+    # Publish multiple events using a generator
+    @publish("my-topic")
+    async def my_generator():
+        for i in range(3):
+            yield {"event_id": i}
+    ```
+    """
+
+    def wrap_coroutine(coro):
+        async def wrapper(*args, **kwargs):
+            # Create the Ensign client if not already created
+            global _client
+            if _client is None:
+                raise RuntimeError(
+                    "publish requires a connection to Ensign, please use the authenticate decorator to provide your credentials"
+                )
```

Comment:
If authenticate has not been called, I thought about having this be the default client, so it uses credentials from the environment. But I'm not sure if that behavior would be too confusing for users.